[JSC] Optimize Array#join's toString operation for Objects

Constellation opened the pull request:
#### 9d0a972fc0bfc1429382127b4ee62cbc5a7bf151
<pre>
[JSC] Optimize Array#join&apos;s toString operation for Objects
<a href="https://bugs.webkit.org/show_bug.cgi?id=255638">https://bugs.webkit.org/show_bug.cgi?id=255638</a>
rdar://108239715

Reviewed by Mark Lam.

This patch optimizes Array#join via several optimizations.

1. Borrow V8&apos;s idea. We cache the last string for Array#join, and if it is identical,
   then we just count it instead of appending it to the vector. Like, &quot;[object Object]&quot; x 100.
   Then we can keep JSStringJoiner&apos;s buffer in a reasonable size. We pre-allocate 16 for Vector
   capacity and which can cover many cases with this without Vector allocation and deallocation.
2. Our &quot;[object Object]&quot; tag string generation is too generic. We should have spec-defined ones
   in SmallStrings, and use JSString from that. This also helps (1) since now &quot;[object Object]&quot;
   will get the exact same JSString.
3. With (1), JSStringJoiner now knows the last JSString*. So if Array#join only sees one element,
   we can just return this string directly.
4. This is the last optimization having the largest effect in the 4. We revisit JSValue::toString
   and JSStringJoiner&apos;s toString operation, and optimize the fast path for Object#toString cases.
   We add JSCell::toStringInline which can handle the cached cases quickly. This also cleans up
   JSValue::toStringSlowCase by extracting JSCell related things to JSCell::toStringInline.

                                  ToT                     Patched

    array-join-object       28.6489+-0.0592     ^     16.4209+-0.1264        ^ definitely 1.7447x faster

* JSTests/microbenchmarks/array-join-object.js: Added.
* Source/JavaScriptCore/runtime/ArrayPrototype.cpp:
(JSC::fastJoin):
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/JSCJSValue.cpp:
(JSC::JSValue::toStringSlowCase const):
(): Deleted.
* Source/JavaScriptCore/runtime/JSCast.h:
* Source/JavaScriptCore/runtime/JSCell.cpp:
(JSC::JSCell::toStringSlowCase const):
* Source/JavaScriptCore/runtime/JSCell.h:
* Source/JavaScriptCore/runtime/JSCellInlines.h:
(JSC::JSCell::toStringInline const):
* Source/JavaScriptCore/runtime/JSGenericTypedArrayViewPrototypeFunctions.h:
(JSC::genericTypedArrayViewProtoFuncJoin):
* Source/JavaScriptCore/runtime/JSObject.cpp:
(JSC::JSObject::toString const):
* Source/JavaScriptCore/runtime/JSString.h:
* Source/JavaScriptCore/runtime/JSStringJoiner.cpp:
(JSC::joinStrings):
(JSC::JSStringJoiner::joinedLength const):
(JSC::JSStringJoiner::~JSStringJoiner): Deleted.
* Source/JavaScriptCore/runtime/JSStringJoiner.h:
(JSC::JSStringJoiner::JSStringJoiner):
(JSC::JSStringJoiner::reserveCapacity):
(JSC::JSStringJoiner::join):
(JSC::JSStringJoiner::append):
(JSC::JSStringJoiner::append8Bit):
(JSC::JSStringJoiner::appendEmptyString):
(JSC::JSStringJoiner::appendWithoutSideEffects):
* Source/JavaScriptCore/runtime/ObjectPrototypeInlines.h:
(JSC::objectPrototypeToStringSlow):
(JSC::objectPrototypeToString):
(JSC::inferBuiltinTag): Deleted.
* Source/JavaScriptCore/runtime/SmallStrings.cpp:
(JSC::SmallStrings::initializeCommonStrings):
(JSC::SmallStrings::visitStrongReferences):
* Source/JavaScriptCore/runtime/SmallStrings.h:
(JSC::SmallStrings::objectNullString const):
(JSC::SmallStrings::objectUndefinedString const):
(JSC::SmallStrings::objectObjectString const):
(JSC::SmallStrings::objectArrayString const):
(JSC::SmallStrings::objectFunctionString const):
(JSC::SmallStrings::objectArgumentsString const):
(JSC::SmallStrings::objectDateString const):
(JSC::SmallStrings::objectRegExpString const):
(JSC::SmallStrings::objectErrorString const):
(JSC::SmallStrings::objectBooleanString const):
(JSC::SmallStrings::objectNumberString const):
(JSC::SmallStrings::objectStringString const):
(JSC::SmallStrings::nullObjectString const): Deleted.
(JSC::SmallStrings::undefinedObjectString const): Deleted.
* Source/WTF/wtf/text/StringView.h:
(WTF::StringViewWithUnderlyingString::StringViewWithUnderlyingString):

Canonical link: <a href="https://commits.webkit.org/263117@main">https://commits.webkit.org/263117@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/79b13bd811a4a362574ecbae1927979f679bba2b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/3665 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/3749 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/3855 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/5091 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/3950 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/3824 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/3754 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/3177 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/3709 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/3939 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/3301 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/4920 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/1445 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/3272 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/3338 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/3001 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/3253 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/3331 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/4693 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/3440 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/3718 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/3007 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/3737 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/3252 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/3272 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/947 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/3284 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/3829 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/419 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/3529 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/1058 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

Jarred-Sumner commented on Source/JavaScriptCore/runtime/JSObject.cpp:
```diff
@@ -2669,8 +2669,14 @@ JSString* JSObject::toString(JSGlobalObject* globalObject) const
+
+    JSValue primitive = callToPrimitiveFunction<CachedSpecialPropertyKey::ToPrimitive>(globalObject, this, vm.propertyNames->toPrimitiveSymbol, PreferString);
     RETURN_IF_EXCEPTION(scope, jsEmptyString(vm));
+    if (!primitive) {
```
wonder if a `LIKELY` here would be helpful

Constellation replied:
Sounds good. Added.

MenloDorian commented on Source/JavaScriptCore/runtime/ArrayPrototype.cpp:
```diff
@@ -400,11 +403,9 @@ inline JSValue fastJoin(JSGlobalObject* globalObject, JSObject* thisObject, Stri
         auto& butterfly = *thisObject->butterfly();
         if (UNLIKELY(length > butterfly.publicLength()))
             break;
-        JSStringJoiner joiner(globalObject, separator, length);
```
Is there a reason why we don't also `reserveCapacity` here?

Constellation replied:
Ditto.

MenloDorian commented on Source/JavaScriptCore/runtime/ArrayPrototype.cpp:
```diff
@@ -577,9 +576,7 @@ JSC_DEFINE_HOST_FUNCTION(arrayProtoFuncToString, (JSGlobalObject* globalObject,
-    JSStringJoiner joiner(globalObject, ","_s, length);
-    RETURN_IF_EXCEPTION(scope, encodedJSValue());
-
+    JSStringJoiner joiner(","_s);
```
Same here: what's the reason for not doing `reserveCapacity`?

Constellation replied:
Because Object case can have duplicate JSString in high possibility. So reserveCapacity will end up wasting much.
On the other hand, double / int32 case, will use this buffer pretty much well.

MenloDorian replied:
Oh right, the object case likely leads to `[Object object]` whereas numbers can have different values.  Makes sense.

MenloDorian reviewed (approved):
r=me

Constellation committed:
[JSC] Optimize Array#join's toString operation for Objects
https://bugs.webkit.org/show_bug.cgi?id=255638
rdar://108239715

Reviewed by Mark Lam.

This patch optimizes Array#join via several optimizations.

1. Borrow V8's idea. We cache the last string for Array#join, and if it is identical,
   then we just count it instead of appending it to the vector. Like, "[object Object]" x 100.
   Then we can keep JSStringJoiner's buffer in a reasonable size. We pre-allocate 16 for Vector
   capacity and which can cover many cases with this without Vector allocation and deallocation.
2. Our "[object Object]" tag string generation is too generic. We should have spec-defined ones
   in SmallStrings, and use JSString from that. This also helps (1) since now "[object Object]"
   will get the exact same JSString.
3. With (1), JSStringJoiner now knows the last JSString*. So if Array#join only sees one element,
   we can just return this string directly.
4. This is the last optimization having the largest effect in the 4. We revisit JSValue::toString
   and JSStringJoiner's toString operation, and optimize the fast path for Object#toString cases.
   We add JSCell::toStringInline which can handle the cached cases quickly. This also cleans up
   JSValue::toStringSlowCase by extracting JSCell related things to JSCell::toStringInline.

                                  ToT                     Patched

    array-join-object       28.6489+-0.0592     ^     16.4209+-0.1264        ^ definitely 1.7447x faster

* JSTests/microbenchmarks/array-join-object.js: Added.
* Source/JavaScriptCore/runtime/ArrayPrototype.cpp:
(JSC::fastJoin):
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/JSCJSValue.cpp:
(JSC::JSValue::toStringSlowCase const):
(): Deleted.
* Source/JavaScriptCore/runtime/JSCast.h:
* Source/JavaScriptCore/runtime/JSCell.cpp:
(JSC::JSCell::toStringSlowCase const):
* Source/JavaScriptCore/runtime/JSCell.h:
* Source/JavaScriptCore/runtime/JSCellInlines.h:
(JSC::JSCell::toStringInline const):
* Source/JavaScriptCore/runtime/JSGenericTypedArrayViewPrototypeFunctions.h:
(JSC::genericTypedArrayViewProtoFuncJoin):
* Source/JavaScriptCore/runtime/JSObject.cpp:
(JSC::JSObject::toString const):
* Source/JavaScriptCore/runtime/JSString.h:
* Source/JavaScriptCore/runtime/JSStringJoiner.cpp:
(JSC::joinStrings):
(JSC::JSStringJoiner::joinedLength const):
(JSC::JSStringJoiner::~JSStringJoiner): Deleted.
* Source/JavaScriptCore/runtime/JSStringJoiner.h:
(JSC::JSStringJoiner::JSStringJoiner):
(JSC::JSStringJoiner::reserveCapacity):
(JSC::JSStringJoiner::join):
(JSC::JSStringJoiner::append):
(JSC::JSStringJoiner::append8Bit):
(JSC::JSStringJoiner::appendEmptyString):
(JSC::JSStringJoiner::appendWithoutSideEffects):
* Source/JavaScriptCore/runtime/ObjectPrototypeInlines.h:
(JSC::objectPrototypeToStringSlow):
(JSC::objectPrototypeToString):
(JSC::inferBuiltinTag): Deleted.
* Source/JavaScriptCore/runtime/SmallStrings.cpp:
(JSC::SmallStrings::initializeCommonStrings):
(JSC::SmallStrings::visitStrongReferences):
* Source/JavaScriptCore/runtime/SmallStrings.h:
(JSC::SmallStrings::objectNullString const):
(JSC::SmallStrings::objectUndefinedString const):
(JSC::SmallStrings::objectObjectString const):
(JSC::SmallStrings::objectArrayString const):
(JSC::SmallStrings::objectFunctionString const):
(JSC::SmallStrings::objectArgumentsString const):
(JSC::SmallStrings::objectDateString const):
(JSC::SmallStrings::objectRegExpString const):
(JSC::SmallStrings::objectErrorString const):
(JSC::SmallStrings::objectBooleanString const):
(JSC::SmallStrings::objectNumberString const):
(JSC::SmallStrings::objectStringString const):
(JSC::SmallStrings::nullObjectString const): Deleted.
(JSC::SmallStrings::undefinedObjectString const): Deleted.
* Source/WTF/wtf/text/StringView.h:
(WTF::StringViewWithUnderlyingString::StringViewWithUnderlyingString):

Canonical link: https://commits.webkit.org/263117@main
